### PR TITLE
Generalize sequence-preserving PyTorch collation

### LINKED ISSUE #254: Refactor and generalize PyTorch collate function to preserve sample sequences and leverage default_collate
**Context & Problem**

In Goldener, intermediate and final results are stored in Pixeltable Tables, which can be converted into PyTorch Datasets. Currently, primitive types and arrays from these datasets may be returned as native Python types (int, np.ndarray) rather than torch.Tensor.

To handle batching, we previously introduced pxt_torch_dataset_collate_fn. However, we have identified two key issues:

- Over-specific design: We initially assumed this collate behavior was only needed for datasets coming directly from Pixeltable. In reality, it is useful for any PyTorch Dataset that formats sample items as a dictionary ready for Goldener tools.

- Standard PyTorch collate_fn list splitting: PyTorch's native collation treats items inside a list as individual features to collate across samples. For example:

Input: [{"label": [class_1, class_2]}, {"label": [class_1, class_2]}]
Standard PyTorch output: batch["label"] = [('class_1', 'class_1'), ('class_2', 'class_2')]
Expected Goldener output: batch["label"] = [('class_1', 'class_2'), ('class_1', 'class_2')]

Additionally, PyTorch raises an error if list lengths vary across samples because it attempts element-wise collation. In Goldener, we operate under the strong assumption that each dataset sample is a dict, and each key-value pair in that dict represents a single complete feature. Sequences should be preserved per sample, leaving tensor conversion to downstream tools when necessary.

**Proposed Solution**

Move the collate function to goldener/utils/torch_utils.py (or the corresponding utility module) and rename it to collate_keeping_sequences_as_sequences to reflect its general utility.

Refactor the logic to delegate non-sequence feature collation directly to PyTorch's standard torch.utils.data.default_collate.
 
**Suggested Implementation**

```
def collate_keeping_sequences_as_sequences(batch: list[dict[str, Any]])  -> dict[str, torch.Tensor | Sequence]:
    value_list_dict = defaultdict(list)
    conversion_dict: dict[str, Callable[[Sequence[Any]], torch.Tensor]] = {}

    for idx_sample, sample in enumerate(batch):
        for key, value in sample.items():
            value_list_dict[key].append(value)
            if idx_sample == 0:
                if not isinstance(value, Sequence):
                    conversion_dict[key] = default_collate

    return {
        key: conversion_dict[key](value) if key in conversion_dict else value
        for key, value in value_list_dict.items()
    }
```

### PR DESCRIPTION
**Pull request by jason-scheffel**
Closes #254.

This moves the sequence-preserving collator to `torch_utils`, renames it
to
    `collate_keeping_sequences_as_sequences`, and delegates non-sequence
fields to PyTorch's `default_collate`.
All existing call sites and GoldDoer defaults use the generalized helper.

### CODE REVIEW

File: goldener/torch_utils.py [hunk 1]
```diff
@@ -1,13 +1,42 @@
...
+            values_by_key[key].append(value)
+
+    return {
+        key: (
+            values
+            if isinstance(values[0], Sequence) or values[0] is None
+            else default_collate(values)
+        )
+        for key, values in values_by_key.items()
+    }
```
[the remark this answers is not in the document]
Comment (jason-scheffel) [reply goldener/torch_utils.py:50]: I verified in the project's Python 3.13 environment: neither `torch.Tensor` nor `np.ndarray` is an instance of `collections.abc.Sequence`. Both therefore reach `default_collate` and are stacked correctly, as covered by the tests. The implementation suggested in #254 uses the same `Sequence` check.

File: goldener/describe.py [hunk 2]
```diff
@@ -56,7 +58,7 @@ class GoldDescriptor:
             applied by the `collate_fn`.
         collate_fn: Optional function to collate dataset samples into batches composed of
             dictionaries with at least the key specified by `data_key` returning a PyTorch Tensor.
-            If None, `pxt_torch_dataset_collate_fn` is used, which preserves list-valued
+            If None, `collate_keeping_sequences_as_sequences` is used, which preserves sequence-valued
             fields as one list per sample. It should format
```
Comment (Yann-CV) [inline goldener/describe.py]:
```suggestion
            fields as one sequence per sample. It should format
```

File: goldener/select.py [hunk 3]
```diff
@@ -652,7 +654,7 @@ class GoldSelector:
         chunk: Optional chunk size for processing data in chunks to reduce memory consumption.
         collate_fn: Optional function to collate dataset samples into batches composed of
             dictionaries with at least the key specified by `vectorized_key` returning a PyTorch Tensor.
-            If None, `pxt_torch_dataset_collate_fn` is used, which preserves list-valued
+            If None, `collate_keeping_sequences_as_sequences` is used, which preserves sequence-valued
             fields as one list per sample.
```
Comment (Yann-CV) [inline goldener/select.py]:
```suggestion
            fields as one sequence per sample.
```

File: goldener/torch_utils.py [hunk 4]
```diff
@@ -1,13 +1,42 @@
...
+        batch: A list of samples, where each sample is a dictionary.
+
+    Returns:
+        A dictionary with sequences kept per sample and all other values collated
+        with PyTorch's default collate function.
+    """
+    values_by_key: dict[str, list[Any]] = defaultdict(list)
+    for sample in batch:
+        for key, value in sample.items():
+            values_by_key[key].append(value)
```
[the remark this answers is not in the document]
Comment (Yann-CV) [reply goldener/torch_utils.py]: not a bad comment!

We should handle this! 

torch is taking the keys from the first element in the batch and iterate over the batch for each key. Probably faster (probably negligible though), to:
1. create `values_by_key` from first sample
 2. iterate over the remaining samples and add values by iterating over the found keys (missing key will raise an error)
 
it will be equifeature with torch. In addition we can as well check if there is extra keys in a sample and log a warning

File: goldener/vectorize.py [hunk 5]
```diff
@@ -474,7 +477,7 @@ class GoldVectorizer:
         vectorizer: GoldTensorVectorizationTool instance for transforming batched inputs into vectors.
         collate_fn: Optional function to collate dataset samples into batches composed of
             dictionaries with at least the key specified by `data_key` returning a PyTorch Tensor.
-            If None, `pxt_torch_dataset_collate_fn` is used, which preserves list-valued
+            If None, `collate_keeping_sequences_as_sequences` is used, which preserves sequence-valued
             fields as one list per sample.
```
Comment (Yann-CV) [inline goldener/vectorize.py]:
```suggestion
            fields as one sequence per sample.
```

File: tests/test_split.py [hunk 6]
```diff
@@ -98,9 +98,9 @@ def test_get_split_indices_from_dataset_with_default_collate_fn(self):
         splitted = GoldSplitter.get_split_indices(
             DummyDataset(
                 [
+                    {"idx": 2, "selected": None},
```
Comment (Yann-CV) [inline tests/test_split.py:101]: ?

File: tests/test_torch_utils.py [hunk 7]
```diff
@@ -13,6 +16,64 @@
 )
 
 
+class TestCollateKeepingSequencesAsSequences:
+    def test_collates_numpy_arrays_and_integers(self):
```
Comment (Yann-CV) [inline tests/test_torch_utils.py]: we can trust pytorch on this

File: tests/test_torch_utils.py [hunk 8]
```diff
@@ -13,6 +16,64 @@
...
+        result = collate_keeping_sequences_as_sequences(batch)
+
+        assert isinstance(result["image"], torch.Tensor)
+        assert isinstance(result["label"], torch.Tensor)
+        assert result["text"] == ["hello", "world"]
+        assert result["image"].shape == (2, 3)
+        assert result["label"].dtype == torch.int64
+        assert torch.equal(result["label"], torch.tensor([0, 1], dtype=torch.int64))
+
+    def test_collates_tensors(self):
```
Comment (Yann-CV) [inline tests/test_torch_utils.py]: this one as well (I guess the one below is enough)

**Review by Yann-CV — CHANGES_REQUESTED** [review]
Great! Thanks for your contribution.

Only, the issue raised by Cubic worth to be handled.

**Comment by Yann-CV** [comment]
@jason-scheffel I added you as contributor here! you can now assign yourself to issues (security against duplicated work)

File: tests/test_split.py (hunk 6 above)
Comment (jason-scheffel) [reply tests/test_split.py:101]: I moved the unassigned sample first because this call omits `collate_fn`, so `get_split_indices` uses `collate_keeping_sequences_as_sequences`. With `None` first, the test exercises its `values[0] is None` guard. Without that guard, the collator would pass `[None, "train", "val"]` to `default_collate`, which raises before `get_split_indices` can skip the unassigned entry.

File: goldener/torch_utils.py [hunk 9]
```diff
@@ -1,13 +1,57 @@
...
+    for sample_idx, sample in enumerate(batch[1:], start=1):
+        for key in values_by_key:
+            values_by_key[key].append(sample[key])
+
+        extra_keys = sample.keys() - values_by_key.keys()
+        if extra_keys:
+            logger.warning(
+                "Ignoring extra keys %s in batch sample at index %d.",
+                sorted(extra_keys),
+                sample_idx,
```
Comment (Yann-CV) [inline goldener/torch_utils.py]:
```suggestion
                f"Ignoring extra keys {sorted(extra_keys)} in batch sample at index {sample_idx}.",
```

Comment (Yann-CV) [reply goldener/torch_utils.py]: f-str is more pythonic

**Review by Yann-CV — APPROVED** [review]
Great! Thanks just commited a modification for the str in the warning.

Hope to review your work again soon!
